src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -592,19 +630,34 @@ public void Add(string key, ModelStateEntry value)
                 throw new ArgumentNullException(nameof(value));
             }
 
-            _innerDictionary.Add(key, value);
+            if (ContainsKey(key))
+            {
+                throw new ArgumentException(Resources.ModelStateDictionary_DuplicateKey, nameof(key));
+            }
+
+            this[key] = value;
         }
 
         /// <inheritdoc />
         public void Clear()
         {
-            _innerDictionary.Clear();
+            Count = 0;
+            HasRecordedMaxModelError = false;
+            ErrorCount = 0;
+            _root.Reset();
+            _root.Children.Clear();
         }
 
         /// <inheritdoc />
-        public bool Contains(KeyValuePair<string, ModelStateEntry> item)
+        bool ICollection<KeyValuePair<string, ModelStateEntry>>.Contains(KeyValuePair<string, ModelStateEntry> item)
```

Comment:
Not sure how to go about this - the equality comparison for `item.Value` will always return false since we create a copy of the one provided to us in Add or indexer. Should we just have this return false?